// react-native-scripts/build/scripts/init.js
'use strict';

var _regenerator = require('babel-runtime/regenerator');

var _regenerator2 = _interopRequireDefault(_regenerator);

var _stringify = require('babel-runtime/core-js/json/stringify');

var _stringify2 = _interopRequireDefault(_stringify);

var _assign = require('babel-runtime/core-js/object/assign');

var _assign2 = _interopRequireDefault(_assign);

var _asyncToGenerator2 = require('babel-runtime/helpers/asyncToGenerator');

var _asyncToGenerator3 = _interopRequireDefault(_asyncToGenerator2);

var _chalk = require('chalk');

var _chalk2 = _interopRequireDefault(_chalk);

var _fsExtra = require('fs-extra');

var _fsExtra2 = _interopRequireDefault(_fsExtra);

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _pathExists = require('path-exists');

var _pathExists2 = _interopRequireDefault(_pathExists);

var _crossSpawn = require('cross-spawn');

var _crossSpawn2 = _interopRequireDefault(_crossSpawn);

var _minimist = require('minimist');

var _minimist2 = _interopRequireDefault(_minimist);

var _log = require('../util/log');

var _log2 = _interopRequireDefault(_log);

var _install = require('../util/install');

var _install2 = _interopRequireDefault(_install);

var _pm = require('../util/pm');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// UPDATE DEPENDENCY VERSIONS HERE
var DEFAULT_DEPENDENCIES = {
  expo: '^27.0.1',
  react: '16.3.1',
  'react-native': '~0.55.2'
};

var WEB_DEFAULT_DEPENDENCIES = {
  'expo-web': '^0.0.12',
  'react-dom': '16.0.0',
  'react-native-web': '^0.4.0',
  webpack: '^3.11.0',
  'webpack-dev-server': '2.9.4'
};

// TODO figure out how this interacts with ejection
var DEFAULT_DEV_DEPENDENCIES = {
  'jest-expo': '~27.0.0',
  'react-test-renderer': '16.3.1'
};

var WEB_DEFAULT_DEV_DEPENDENCIES = {
  'react-native-scripts': 'latest',
  'babel-loader': '^7.1.2',
  'babel-plugin-expo-web': '^0.0.5',
  'babel-plugin-react-native-web': '^0.4.0',
  'babel-plugin-transform-decorators-legacy': '^1.3.4',
  'babel-plugin-transform-imports': '^1.4.1',
  'babel-plugin-transform-runtime': '^6.23.0',
  'file-loader': '^1.1.7',
  'css-loader': '^0.28.7',
  'style-loader': '^0.19.0'
};

var arg = (0, _minimist2.default)(process.argv.slice(2), {
  boolean: ['with-web-support']
});

module.exports = function () {
  var _ref = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee(appPath, appName, verbose) {
    var cwd = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : '';

    var ownPackageName, ownPath, useYarn, npmOrYarn, npmVersion, npmVersionParts, majorVersion, minorVersion, patchVersion, readmeExists, appPackagePath, appPackage, withWebSupport, data, _ref2, code, command, args, cdpath;

    return _regenerator2.default.wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            ownPackageName = require('../../package.json').name;
            ownPath = _path2.default.join(appPath, 'node_modules', ownPackageName);
            useYarn = (0, _pm.hasYarn)(appPath);
            npmOrYarn = useYarn ? 'yarn' : 'npm';


            if (!useYarn) {
              npmVersion = _crossSpawn2.default.sync('npm', ['--version']).stdout.toString().trim();
              npmVersionParts = npmVersion.split('.');
              majorVersion = parseInt(npmVersion[0], 10);
              minorVersion = parseInt(npmVersion[1], 10);
              patchVersion = parseInt(npmVersion[2], 10);


              if (majorVersion === 5 && minorVersion < 7) {
                console.log(_chalk2.default.yellow('\n*******************************************************************************\nERROR: npm >= 5.0.0 and < 5.7.0 are not supported\n*******************************************************************************\n\nIt looks like you\'re using a version of npm that is buggy with this tool.\n\nWe recommend using npm >= 5.7.0 or yarn.\n\n*******************************************************************************\n '));
                process.exit(1);
              }
            }

            _context.next = 7;
            return (0, _pathExists2.default)(_path2.default.join(appPath, 'README.md'));

          case 7:
            readmeExists = _context.sent;

            if (!readmeExists) {
              _context.next = 11;
              break;
            }

            _context.next = 11;
            return _fsExtra2.default.rename(_path2.default.join(appPath, 'README.md'), _path2.default.join(appPath, 'README.old.md'));

          case 11:
            appPackagePath = _path2.default.join(appPath, 'package.json');
            _context.t0 = JSON;
            _context.next = 15;
            return _fsExtra2.default.readFile(appPackagePath);

          case 15:
            _context.t1 = _context.sent;
            appPackage = _context.t0.parse.call(_context.t0, _context.t1);


            // mutate the default package.json in any ways we need to
            appPackage.main = './node_modules/react-native-scripts/build/bin/crna-entry.js';
            appPackage.scripts = {
              start: 'react-native-scripts start',
              eject: 'react-native-scripts eject',
              android: 'react-native-scripts android',
              ios: 'react-native-scripts ios',
              test: 'jest'
            };

            withWebSupport = arg['with-web-support'];

            if (withWebSupport) {
              appPackage.main = './node_modules/react-native-scripts/build/bin/crna-entry-web.js';
              (0, _assign2.default)(appPackage.scripts, {
                web: 'webpack-dev-server -d --config ./webpack.config.js  --inline --hot --colors --content-base public/ --history-api-fallback',
                build: 'NODE_ENV=production webpack -p --config ./webpack.config.js'
              });
            }

            appPackage.jest = {
              preset: 'jest-expo'
            };

            if (!appPackage.dependencies) {
              appPackage.dependencies = {};
            }

            if (!appPackage.devDependencies) {
              appPackage.devDependencies = {};
            }

            // react-native-scripts is already in the package.json devDependencies
            // so we need to merge instead of assigning
            (0, _assign2.default)(appPackage.dependencies, DEFAULT_DEPENDENCIES);
            (0, _assign2.default)(appPackage.devDependencies, DEFAULT_DEV_DEPENDENCIES);

            if (withWebSupport) {
              (0, _assign2.default)(appPackage.dependencies, WEB_DEFAULT_DEPENDENCIES);
              (0, _assign2.default)(appPackage.devDependencies, WEB_DEFAULT_DEV_DEPENDENCIES);
            }

            // Write the new appPackage after copying so that we can include any existing
            _context.next = 29;
            return _fsExtra2.default.writeFile(appPackagePath, (0, _stringify2.default)(appPackage, null, 2));

          case 29:
            _context.next = 31;
            return _fsExtra2.default.copy(_path2.default.join(ownPath, arg['with-web-support'] ? 'template-with-web' : 'template'), appPath);

          case 31:
            _context.prev = 31;
            _context.next = 34;
            return _fsExtra2.default.rename(_path2.default.join(appPath, 'gitignore'), _path2.default.join(appPath, '.gitignore'));

          case 34:
            _context.next = 49;
            break;

          case 36:
            _context.prev = 36;
            _context.t2 = _context['catch'](31);

            if (!(_context.t2.code === 'EEXIST')) {
              _context.next = 48;
              break;
            }

            _context.next = 41;
            return _fsExtra2.default.readFile(_path2.default.join(appPath, 'gitignore'));

          case 41:
            data = _context.sent;
            _context.next = 44;
            return _fsExtra2.default.appendFile(_path2.default.join(appPath, '.gitignore'), data);

          case 44:
            _context.next = 46;
            return _fsExtra2.default.unlink(_path2.default.join(appPath, 'gitignore'));

          case 46:
            _context.next = 49;
            break;

          case 48:
            throw _context.t2;

          case 49:
            _context.next = 51;
            return (0, _install2.default)(appPath);

          case 51:
            _ref2 = _context.sent;
            code = _ref2.code;
            command = _ref2.command;
            args = _ref2.args;

            if (!(code !== 0)) {
              _context.next = 58;
              break;
            }

            console.error('Failed to install');
            // console.error(`\`${command} ${args.join(' ')}\` failed`);
            return _context.abrupt('return');

          case 58:

            // display the cleanest way to get to the app dir
            // if the cwd + appName is equal to the full path, then just cd into appName
            cdpath = void 0;

            if (_path2.default.resolve(cwd, appName) === appPath) {
              cdpath = appName;
            } else {
              cdpath = appPath;
            }

            (0, _log2.default)('\nSuccess! Created ' + appName + ' at ' + appPath + '\nInside that directory, you can run several commands:\n\n  ' + _chalk2.default.cyan(npmOrYarn + ' start') + '\n    Starts the development server so you can open your app in the Expo\n    app on your phone.\n\n  ' + _chalk2.default.cyan(npmOrYarn + ' run ios') + '\n    (Mac only, requires Xcode)\n    Starts the development server and loads your app in an iOS simulator.\n\n  ' + _chalk2.default.cyan(npmOrYarn + ' run android') + '\n    (Requires Android build tools)\n    Starts the development server and loads your app on a connected Android\n    device or emulator.\n  ' + (withWebSupport ? webLogMessage(npmOrYarn) : '\n') + '\n  ' + _chalk2.default.cyan(npmOrYarn + ' test') + '\n    Starts the test runner.\n\n  ' + _chalk2.default.cyan(npmOrYarn + ' run eject') + '\n    Removes this tool and copies build dependencies, configuration files\n    and scripts into the app directory. If you do this, you can\u2019t go back!\n\n\nWe suggest that you begin by typing:\n\n  ' + _chalk2.default.cyan('cd ' + cdpath) + '\n  ' + _chalk2.default.cyan(npmOrYarn + ' start'));

            if (readmeExists) {
              (0, _log2.default)('\n' + _chalk2.default.yellow('You had a `README.md` file, we renamed it to `README.old.md`'));
            }

            (0, _log2.default)();
            (0, _log2.default)('Happy hacking!');

          case 64:
          case 'end':
            return _context.stop();
        }
      }
    }, _callee, undefined, [[31, 36]]);
  }));

  return function (_x, _x2, _x3) {
    return _ref.apply(this, arguments);
  };
}();

function webLogMessage(npmOrYarn) {
  return '\n  ' + _chalk2.default.cyan(npmOrYarn + ' web') + '\n    Starts the Webpack server to serve the web version of the app.\n  ';
}
